### PR TITLE
SCMOD-8463: Run health checks before starting worker

### PR DESCRIPTION
**Pull request by kusumaghoshdastidar**
Jira: https://portal.digitalsafe.net/browse/SCMOD-8463

Dev builds: http://sou-jenkins2.swinfra.net/job/WorkerFramework/view/Developer/job/WorkerFramework%7Eworker-framework%7ESCMOD-8463%7ECI/

### CODE REVIEW

File: worker-archetype/src/main/resources/archetype-resources/__rootArtifactId__-container/pom.xml [hunk 1]
```diff
@@ -34,7 +34,6 @@
     <properties>
         <maven.install.skip>true</maven.install.skip>
         <maven.deploy.skip>true</maven.deploy.skip>
-        <docker.maven.version>0.13.6</docker.maven.version>
```
Comment (dermot-hardy) [inline worker-archetype/src/main/resources/archetype-resources/__rootArtifactId__-container/pom.xml:37]: Are these archetype updates/fixes required for this Jira?  If not then remove them from this PR and put them in a separate one.

File: worker-core/src/main/java/com/hpe/caf/worker/core/GatedHealthProvider.java [hunk 2]
```diff
@@ -57,7 +59,6 @@ protected Result check() throws Exception
             synchronized (healthCheckLock) {
 
                 final Result result = healthCheck.execute();
-
```
Comment (dermot-hardy) [inline worker-core/src/main/java/com/hpe/caf/worker/core/GatedHealthProvider.java]: Pointless whitespace change.

File: worker-core/src/main/java/com/hpe/caf/worker/core/WorkerCore.java [hunk 3]
```diff
@@ -61,7 +63,23 @@ public WorkerCore(final Codec codec, final WorkerThreadPool pool, final ManagedW
...
     {
-        workerQueue.start(callback);
+        try
+        {
+            workerQueue.start(callback);
+            isStarted = true;
+        }
+        catch(final QueueException e) {
+            throw e;
+        }
```
Comment (dermot-hardy) [inline worker-core/src/main/java/com/hpe/caf/worker/core/WorkerCore.java]: The try-catch seems pointless.

File: worker-core/src/main/java/com/hpe/caf/worker/core/GatedHealthProvider.java [hunk 4]
```diff
@@ -67,13 +68,20 @@ protected Result check() throws Exception
...
+                    else
+                    {
+                        LOG.debug("Reconnecting the incoming queue as all health checks are passing");
+                        workerQueue.reconnectIncoming();
+                    }
                 } else {
                     LOG.debug("Disconnecting the incoming queue due to the [{}] health check failing", name);
                     workerQueue.disconnectIncoming();
                 }
-
```
Comment (dermot-hardy) [inline worker-core/src/main/java/com/hpe/caf/worker/core/GatedHealthProvider.java]: Pointless whitespace change.

File: worker-core/src/main/java/com/hpe/caf/worker/core/GatedHealthProvider.java [hunk 5]
```diff
@@ -67,13 +68,20 @@ protected Result check() throws Exception
...
+                        core.start();
+                    }
+                    else
+                    {
+                        LOG.debug("Reconnecting the incoming queue as all health checks are passing");
+                        workerQueue.reconnectIncoming();
+                    }
                 } else {
                     LOG.debug("Disconnecting the incoming queue due to the [{}] health check failing", name);
                     workerQueue.disconnectIncoming();
```
Comment (dermot-hardy) [inline worker-core/src/main/java/com/hpe/caf/worker/core/GatedHealthProvider.java:84]: Presumably it is safe to call this even if it `core.start();` was never called?

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
Comments above.

File: worker-archetype/src/main/resources/archetype-resources/__rootArtifactId__-container/pom.xml (hunk 1 above)
Comment (kusumaghoshdastidar) [reply worker-archetype/src/main/resources/archetype-resources/__rootArtifactId__-container/pom.xml:37]: With the change made for this jira, the tests fail without the healthcheck being called. The changes in the archetype are to get the container to call the healthcheck.

File: worker-core/src/main/java/com/hpe/caf/worker/core/GatedHealthProvider.java (hunk 5 above)
Comment (kusumaghoshdastidar) [reply worker-core/src/main/java/com/hpe/caf/worker/core/GatedHealthProvider.java:84]: Seems to be working fine, here is a scenario when worker core was not started:
`[INFO] 18:04:11.294 worker-example> [01:04:11.293Z #2fd.021 INFO  -            -   ] c.h.c.i.w.e.ExampleWorker: Running healthCheck...
[INFO] 18:04:11.294 worker-example> [01:04:11.293Z #2fd.021 DEBUG -            -   ] c.h.c.i.w.e.ExampleWorker: The Example Worker's Health Check is set to Unhealthy
[INFO] 18:04:11.294 worker-example> [01:04:11.293Z #2fd.021 DEBUG -            -   ] c.h.c.w.c.GatedHealthProvider: Disconnecting the incoming queue due to the [worker] health check failing
[INFO] 18:04:11.294 worker-example> [01:04:11.293Z #2fd.021 DEBUG -            -   ] c.h.c.w.q.r.RabbitWorkerQueue: Disconnecting incoming queues
[INFO] 18:04:11.295 worker-example> [01:04:11.295Z #2fd.021 DEBUG -            -   ] c.h.c.w.c.GatedHealthProvider: Disconnecting the incoming queue due to the [queue] health check failing
[INFO] 18:04:11.295 worker-example> [01:04:11.295Z #2fd.021 DEBUG -            -   ] c.h.c.w.q.r.RabbitWorkerQueue: Disconnecting incoming queues
`

File: worker-core/src/main/java/com/hpe/caf/worker/core/WorkerCore.java [hunk 6]
```diff
@@ -61,7 +63,17 @@ public WorkerCore(final Codec codec, final WorkerThreadPool pool, final ManagedW
     public void start()
         throws QueueException
     {
-        workerQueue.start(callback);
+         workerQueue.start(callback);
+         isStarted = true;
```
Comment (dermot-hardy) [inline worker-core/src/main/java/com/hpe/caf/worker/core/WorkerCore.java]: The indenting is off here - 5 spaces instead of 4.

**Review by dermot-hardy — APPROVED** [review]

**Pull request merged**